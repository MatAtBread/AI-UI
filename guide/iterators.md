# Iterators

As you can tell, AIUI makes a lot of use of Async Iterators (hence the name: "AI-UI"!).

Later, we'll make a lot of use of async iterators created by AI-UI from standard DOM elements and events (see [Link elements dynamically](./when.md)), but before we do, you should familiarise yourself with a few general helpers provided by AI-UI for working with async iterables.

First, some clarity on terminology.
* An **async iterator** is an object that resolves multiple times in the future. It has optional methods that allow the _consumer_ of these values to terminate the thing that is _generating_ them. Unlike Observables which "push" values to a subscriber, an **async iterator** is a "pull" mechanism - nothing is generated until there is a consumer for the values, and both the _consumer_ and _generator_ of the values can terminate the iteration.
* An **async iterable** is an object that can be iterated over asynchronously. 
* An **async generator** is a function that returns an **async iterable** when it is invoked, using the `yield` and `await` Javascript keywords.

At present, none of these types have standardised prototypes, as they are not Javascript classes, but simply objects that have a standard set of interfaces, as indicated by the presence of a various well-known Symbols on objects and corresponding methods. You can find out more about the details [here](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Iteration_protocols#the_async_iterator_and_async_iterable_protocols). There are proposals to expose some of the prototypes so that they can be expanded in the future, for example the [TC39 proposal](https://github.com/tc39/proposal-async-iterator-helpers).

Meanwhile, AI-UI has a set of helpers you can use with async iterators.

To use these with third-party async iterators, or the return of a standard async generator, use the `iterableHelpers` or `generatorHelpers` function, depdending on if you want to add helpers to a generator (that returns iterables), or an iterable itself.

Of these, `map` and `filter` are the most useful with AI-UI, as you can see in the example following the defintions of the helpers. 

```javascript
/* ES6 Import */
import { Iterators } from 'https://www.unpkg.com/@matatbread/ai-ui/esm/ai-ui.js'; 
const { iterableHelpers, generatorHelpers } = Iterators;
```

```javascript
/* CommonJS */
const { Iterators } = require('@matatbread/ai-ui'); 
const { iterableHelpers, generatorHelpers } = Iterators;
```


```html
<script src="https://www.unpkg.com/@matatbread/ai-ui/dist/ai-ui.min.js"></script> 
<script>
  /* Static script */
  const { iterableHelpers, generatorHelpers } = AIUI.Iterators;
</script>
```

These functions manipulate an async iterable or generator, adding helper functions.

For example, consider an async generator that counts from 0 to a specified limit:

```javascript
async function *count(limit) {
  for (let i=0; i<limit; i++)
    yield i;
}
```

We can either add functionality to the generator with `generatorHelpers`, or to the result of calling the generator (the async iterable it returns) with `iterableHelpers`. 

```javascript
// Create a version of "count" whose returned iterators have helpers attached
const helpedCount1 = generatorHelpers(count);
const counter1 = helpedCount(10);

// or just:
const helpedCount2 = generatorHelpers(async function *(limit) {
  for (let i=0; i<limit; i++)
    yield i;
});
const counter2 = helpedCount2(10);

// or call the intrinsic `count` and add helpers to the result:
const counter3 = iterableHelpers(count(10));
```
In each case, the resulting async iterable will function just as a standard async interable:

```javascript
for await (const x of counter3 /* or counter1, or counter2 */) {
  console.log(x); // Integers 0,1,2...9
}
```
..but will now have the following additional propertied:

## map
```typescript
function map<U, R>(this: AsyncIterable<U>, mapper: (o: U) => R | PromiseLike<R>): AsyncIterable<Awaited<R>>
```

Maps the results of an async iterable. The mapper function can itself be asynchronous.

```javascript
for await (const x of counter3.map(num => num * num)) {
  console.log(x); // Integers 0,1,4...81
}
```

## filter
```typescript
function filter<U>(this: AsyncIterable<U>, fn: (o: U) => boolean | PromiseLike<boolean>): AsyncIterable<U>
```

Filter the results of an async iterable. Only those vales returning `true` from the predicate are yielded to the consumer. The predicate can be asynchronous.
```javascript
for await (const x of counter3.filter(num => num % 2 === 0)) {
  console.log(x); // Integers 0,2,4,6,8
}
```

## initially
```typescript
function initially<U, I = U>(this: AsyncIterable<U>, initValue: I): AsyncIterable<U | I>
```
Prepends a single value to a yieded sequence
```javascript
for await (const x of counter3.initially('hello')) {
  console.log(x); // 'hello',0,1,2,3,4,5,6,7,8,9
}
```

## throttle
```typescript
function throttle<U>(this: AsyncIterable<U>, milliseconds: number): AsyncIterable<U>
```

Filters out yielded values which occur within `milliseconds` or the previously yielded value

## waitFor
```typescript
function waitFor<U>(this: AsyncIterable<U>, cb: (done: (value: void | PromiseLike<void>) => void) => void): AsyncIterable<U>
```

Waits for a callback before yielding values as they arrive. For example, to yield on the next animation frame:
```javascript
for await (const x of counter3.waitFor(done => window.requestAnimationFrame(done))) {
  // We've been scheduled to run during an animation frame period
}
```

## consume
```typescript
function consume<U>(this: AsyncIterable<U>, f?: ((u: U) => void | PromiseLike<void>) | undefined): Promise<void>
```

Passes each yielded value to the specified function and returns on when the final iteration has been processed by the callback. The specified function can itself be synchronous, async (returning a Promise), `null` or `undefined`.

```javascript
// Outputs all the values 0..9 and then continues
await counter3.consume(n => console.log(n));
```

## broadcast
```typescript
function broadcast<U, X>(this: AsyncIterable<U>, pipe?: (dest: AsyncIterable<U>) => AsyncIterable<X>): AsyncIterable<X>
```

Queues the incoming yielded values so they can be replayed to multiple consumers. The queue for each consumer commences when the consumer first calls the async iterator `next()` function, and the whole broadcast iterator terminates when either the producer or all consumers terminate.

```javascript
const b = counter3.broadcast();

// Outputs A0,B0,A1,B1.... note: the synchronisation and order between A and B may differ depending on the timing of functions. The order within "A" or "B" will be maintained.

b.consume(n => console.log("A",n));
b.consume(n => console.log("B",n));

```

Note that all the helpers (except consume) themselves return "helped" async iterables, so you can chain them together:
```javascript
for await (const n of counter3.filter(n => n % 2 === 1).map(n => n * n).initially(-1).waitFor(requestAnimationFrame)) {
  console.log(n); // -1,1,9,25,49,81, all output during an animation frame
}
```


In the following example, we use the same techniques as we did in [Dynamic content](./dynamic-content.md), but rather than making the generator return DOM Nodes, we map the raw "data" yielded by the generator into the required UI _within_ the static UI declaration. This helps us with "separation of concerns", where the async iterables generate data, and the UI specifies how it should be presented to the user.

Back to Chuck Norris.

Here's our existing version:

```javascript
const App = div.extended({
  constructed() {
    /* When constructed, this "div" tag contains some other tags: */
    return [
      h2("Hello World"),
      div(chuckJoke())
    ]
  }
});

/* Add add it to the document so the user can see it! */
document.body.appendChild(App({
  style:{
    color: 'blue'
  }
},
'Your first web page'));

/* A simple async "sleep" function */
function sleep(seconds) {
  return new Promise(resolve => setTimeout(resolve, seconds * 1000))
}

async function *chuckJoke() {
  while (true) {
    /* In this example, we use Promises, rather than await, just to illustrate the alternative
    JavaScript syntax */
    yield fetch('https://api.chucknorris.io/jokes/random')
      .then(response => response.json())
      .then(fromJson => fromJson.value);
    await sleep(5);
  }
}
```
What if we wanted the joke in a bigger font? We could just change `chuckJoke` to yield a whole `div`, complete with it's own style, but this blurs the boundary between what is "data" as what is "presentation".

An alternative is to map the joke (text data) to UI elements in the `App`:

```javascript
const App = div.extended({
  constructed() {
    /* When constructed, this "div" tag contains some other tags: */
    return [
      h2("Hello World"),
      chuckJoke().map(joke => div({ style: { fontSize: '133%' }},joke))
    ]
  }
});

```
Of course, to do this, we need to add the helpers to `chuckJoke`. We could do it where it's referenced, like:
```javascript
      iterableHelpers(chuckJoke()).map(joke => div({ style: { fontSize: '133%' }},joke))
```
...but since it's a generator, it's cleaner to do it where it's defined:
```javascript
const chuckJoke = generatorHelpers(async function *() {
  while (true) {
    /* In this example, we use Promises, rather than await, just to illustrate the alternative
    JavaScript syntax */
    yield fetch('https://api.chucknorris.io/jokes/random')
      .then(response => response.json())
      .then(fromJson => fromJson.value);
    await sleep(5);
  }
});

```

Try the above code [example (right click and open in new tab)](https://raw.githack.com/MatAtBread/AI-UI/main/guide/examples/iterators.html)


____

| < Prev | ^ |  Next > |
|:-------|:-:|--------:|
| [Dynamic Content](./dynamic-attributes.md) | [Index](./index.md) | [Create new tags](./extended.md) |



