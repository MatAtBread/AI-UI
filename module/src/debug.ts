// @ts-ignore
export const DEBUG = globalThis.DEBUG == '*' || globalThis.DEBUG == true || globalThis.DEBUG?.match(/(^|\W)AI-UI(\W|$)/) || false;
export function log(...args:any) {
    if (DEBUG) console.log('(AI-UI)', ...args)
}
