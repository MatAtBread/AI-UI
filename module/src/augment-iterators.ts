import { augmentGlobalAsyncGenerators } from './ai-ui.js';
import { AsyncExtraIterable } from './iterators.js';
export * from './iterators.js';

// Find the prototype that holds the [Symbol.asyncIterator] for an async generator function, and add the helpers to it.
augmentGlobalAsyncGenerators();

export {};

declare global {
  interface AsyncGenerator<T = unknown, TReturn = any, TNext = unknown> extends AsyncIterator<T, TReturn, TNext>, AsyncExtraIterable<T> {
  }
}