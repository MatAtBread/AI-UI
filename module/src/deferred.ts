import { DEBUG, console } from "./debug.js";

// Create a deferred Promise, which can be asynchronously/externally resolved or rejected.
export type DeferredPromise<T> = Promise<T> & {
  resolve: (value: T | PromiseLike<T>) => void;
  reject: (value: any) => void;
}

// Used to suppress TS error about use before initialisation
const nothing = (v: any)=>{};

export function deferred<T>(): DeferredPromise<T> {
  let resolve: (value: T | PromiseLike<T>) => void = nothing;
  let reject: (value: any) => void = nothing;
  const promise = new Promise<T>((...r) => [resolve, reject] = r) as DeferredPromise<T>;
  promise.resolve = resolve;
  promise.reject = reject;
  if (DEBUG) {
    const initLocation = new Error().stack;
    promise.catch(ex => (ex instanceof Error || ex?.value instanceof Error) ? console.log("Deferred rejection", ex, "allocated at ", initLocation) : undefined);
  }
  return promise;
}

export function isPromiseLike<T>(x: any): x is PromiseLike<T> {
  return x !== null && x !== undefined && typeof x.then === 'function';
}
