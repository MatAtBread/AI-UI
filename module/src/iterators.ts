import { isAsyncIterable } from "./ai-ui.js";
import { DeferredPromise, deferred } from "./deferred.js";

/* Things to suppliement the JS base AsyncIterable */

export type PushIterator<T> = AsyncExtraIterable<T> & {
  push(value: T): boolean;  // Push a new value to consumers
  close(ex?: Error): void;  // Tell the consumer we're done, with an optional error
};
export type BroadcastIterator<T> = PushIterator<T>;

export interface AsyncExtraIterable<T> extends AsyncIterable<T>, AsyncIterableHelpers {}

/* A function that wraps a "prototypical" AsyncIterator helper, that has `this:AsyncIterable<T>` and returns
  something that's derived from AsyncIterable<R>, result in a wrapped function that accepts
  the same arguments returns a AsyncExtraIterable<X>
*/
function wrapAsyncHelper<T, Args extends any[], R, X extends AsyncIterable<R>>(fn: (this: AsyncIterable<T>, ...args: Args) => X) {
  return function(this: AsyncIterable<T>, ...args:Args) { return withHelpers(fn.call(this, ...args)) }
}

type AsyncIterableHelpers = typeof asyncExtras;
export const asyncExtras = {
  map: wrapAsyncHelper(map),
  filter: wrapAsyncHelper(filter),
  throttle: wrapAsyncHelper(throttle),
  debounce: wrapAsyncHelper(debounce),
  waitFor: wrapAsyncHelper(waitFor),
  count: wrapAsyncHelper(count),
  retain: wrapAsyncHelper(retain),
  broadcast: wrapAsyncHelper(broadcast),
  initially: wrapAsyncHelper(initially),
  consume
};

class QueueIteratableIterator<T> implements AsyncIterableIterator<T> {
  private _pending = [] as DeferredPromise<IteratorYieldResult<T>>[];
  private _items = [] as T[];

  constructor(private readonly stop = () => { }) {
  }

  [Symbol.asyncIterator](this: AsyncIterableIterator<T>) {
    return this;
  }

  next() {
    if (this._items.length) {
      return Promise.resolve({ done: false, value: this._items.shift()! });
    } 
    
    const value = deferred<IteratorYieldResult<T>>();
    this._pending.push(value);
    return value;
  }

  return() {
    const value = { done: true as const, value: undefined };
    if (this._pending) {
      try { this.stop() } catch (ex) { }
      while(this._pending.length)
        this._pending.shift()!.reject(value);
      this._items.splice(0, this._items.length);
    }
    return Promise.resolve(value);
  }

  throw(...args: any[]) {
    const value = { done: true as const, value: args[0] };
    if (this._pending) {
      try { this.stop() } catch (ex) { }
      while(this._pending.length)
        this._pending.shift()!.reject(value);
      this._items.splice(0, this._items.length);
    }
    return Promise.resolve(value);
  }

  push(value: T) {
    if (!this._pending) {
      //throw new Error("pushIterator has stopped");
      return true;
    }
    if (this._pending.length) {
      this._pending.shift()!.resolve({ done: false, value });
    } else {
      this._items.push(value);
    }
    return false;
  }
}

/* An AsyncIterable which typed objects can be published to.
  The queue can be read by multiple consumers, who will each receive
  unique values from the queue (ie: the queue is SHARED not duplicated)
*/
export function pushIterator<T>(stop = () => { }, bufferWhenNoConsumers = false): PushIterator<T> {
  let consumers = 0;
  let ai: QueueIteratableIterator<T> = new QueueIteratableIterator<T>(() => {
    consumers -= 1;
    if (consumers === 0 && !bufferWhenNoConsumers) {
      try { stop() } catch (ex) { }
      // This should never be referenced again, but if it is, it will throw
      (ai as any) = null;
    }
  });

  return Object.assign(Object.create(asyncExtras) as AsyncExtraIterable<T>, {
    [Symbol.asyncIterator]() {
      consumers += 1;
      return ai;
    },
    push(value: T) {
      if (!bufferWhenNoConsumers && consumers === 0) {
        // No one ready to read the results
        return false;
      }
      return ai.push(value);
    },
    close(ex?: Error) {
      ex ? ai.throw?.(ex) : ai.return?.();
      // This should never be referenced again, but if it is, it will throw
      (ai as any) = null;
    }
  });
}

/* An AsyncIterable which typed objects can be published to.
  The queue can be read by multiple consumers, who will each receive
  a copy of the values from the queue (ie: the queue is BROADCAST not shared).

  The iterators stops running when the number of consumers decreases to zero
*/

export function broadcastIterator<T>(stop = () => { }): BroadcastIterator<T> {
  let ai = new Set<QueueIteratableIterator<T>>();

  return <BroadcastIterator<T>>Object.assign(Object.create(asyncExtras) as AsyncExtraIterable<T>, {
    [Symbol.asyncIterator](): AsyncIterableIterator<T> {
      const added = new QueueIteratableIterator<T>(() => {
        ai.delete(added);
        if (ai.size === 0) {
          try { stop() } catch (ex) { }
          // This should never be referenced again, but if it is, it will throw
          (ai as any) = null;
        }
      });
      ai.add(added);
      return withHelpers(added);
    },
    push(value: T) {
      if (!ai?.size)
        return false;

      for (const q of ai.values()) {
        q.push(value);
      }
    },
    close(ex?: Error) {
      for (const q of ai.values()) {
        ex ? q.throw?.(ex) : q.return?.();
      }
      // This should never be referenced again, but if it is, it will throw
      (ai as any) = null;
    }
  });
}

/* Merge asyncIterables into a single asyncIterable */

/* TS hack to expose the return AsyncGenerator a generator of the union of the merged types */
type CollapseIterableType<T> = T[] extends AsyncIterable<infer U>[] ? U : never;
type CollapseIterableTypes<T> = AsyncIterable<CollapseIterableType<T>>;

export const merge = <A extends AsyncIterable<any>[]>(...ai: A) => {
  const it = ai.map(i => Symbol.asyncIterator in i ? i[Symbol.asyncIterator]() : i);
  const promises = it.map((i, idx) => i.next().then(result => ({ idx, result })));
  const results: CollapseIterableType<A[number]>[] = [];
  const forever = new Promise<any>(() => { });
  let count = promises.length;
  const merged: AsyncIterableIterator<A[number]> = {
    [Symbol.asyncIterator]() { return this },
    next() {
      return count 
        ? Promise.race(promises).then(({ idx, result }) => {
          if (result.done) {
            count--;
            promises[idx] = forever;
            results[idx] = result.value;
            return { done: count === 0, value: result.value }
          } else {
            promises[idx] = it[idx].next().then(result => ({ idx, result }));
            return result;
          }
        }).catch(ex => {
          return this.throw?.(ex) ?? Promise.reject({ done: true as const, value: new Error("Iterator merge exception") });
        })
      : Promise.reject({ done: true as const, value: new Error("Iterator merge complete") });
    },
    return() {
      const ex = new Error("Merge terminated");
      for (let i = 0; i < it.length; i++) {
        if (promises[i] !== forever) {
          promises[i] = forever;
          it[i].return?.({ done: true, value: ex }); // Terminate the sources with the appropriate cause
        }
      }
      return Promise.resolve({ done: true, value: ex });
    },
    throw(ex: any) {
      for (let i = 0; i < it.length; i++) {
        if (promises[i] !== forever) {
          promises[i] = forever;
          it[i].throw?.(ex); // Terminate the sources with the appropriate cause
        }
      }
      return Promise.reject(ex);
    }
  };
  return withHelpers(merged as unknown as CollapseIterableTypes<A[number]>);
}


/* 
  Extensions to the AsyncIterable:
  calling `bind(ai)` adds "standard" methods to the specified AsyncIterable
*/

function isExtraIterable<T>(i: any): i is AsyncExtraIterable<T> {
  return isAsyncIterable(i) 
    && Object.keys(asyncExtras)
    .every(k => (k in i) && (i as any)[k] === asyncExtras[k as keyof AsyncIterableHelpers]);
}

// Attach the pre-defined helpers onto an AsyncIterable and return the modified object correctly typed
export function withHelpers<A extends AsyncIterable<any>>(ai: A) {
  if (!isExtraIterable(ai)) {
    Object.assign(ai, asyncExtras);
  }
  return ai as A & (A extends AsyncIterable<infer T> ? AsyncExtraIterable<T> : never)
}

/* AsyncIterable helpers, which can be attached to an AsyncIterator with `withHelpers(ai)`, and invoked directly for foreign asyncIterators */
async function* map<U, R>(this: AsyncIterable<U>, mapper: (o: U) => R | PromiseLike<R>): AsyncIterable<Awaited<R>> {
  const ai = this[Symbol.asyncIterator]();
  try {
    while (true) {
      const p = await ai.next();
      if (p.done) {
        return ai.return?.(mapper(p.value));
      }
      yield mapper(p.value);
    }
  } catch (ex) {
    ai.throw?.(ex);
  }
}

async function* filter<U>(this: AsyncIterable<U>, fn: (o: U) => boolean | PromiseLike<boolean>): AsyncIterable<U> {
  const ai = this[Symbol.asyncIterator]();
  try {
    while (true) {
      const p = await ai.next();
      if (p.done) {
        return ai.return?.(p.value);
      }
      if (await fn(p.value)) {
        yield p.value;
      }
    }
  } catch (ex) {
    ai.throw?.(ex);
  }
}

async function *initially<U, I>(this: AsyncIterable<U>, initValue: I): AsyncIterable<U | I> {
  yield initValue;
  for await (const u of this)
    yield u;
}


async function* throttle<U>(this: AsyncIterable<U>, milliseconds: number): AsyncIterable<U> {
  const ai = this[Symbol.asyncIterator]();
  let paused = 0;
  try {
    while (true) {
      const p = await ai.next();
      if (p.done) {
        return ai.return?.(p.value);
      }
      const now = Date.now();
      if (paused < now) {
        paused = now + milliseconds;
        yield p.value;
      }
    }
  } catch (ex) {
    ai.throw?.(ex);
  }
}

const forever = new Promise<any>(() => { });

// NB: DEBOUNCE IS CURRENTLY BROKEN
async function* debounce<U>(this: AsyncIterable<U>, milliseconds: number): AsyncIterable<U> {
  const ai = this[Symbol.asyncIterator]();
  let timer: Promise<{ debounced: number, value: U }> = forever;
  let last = -1;
  try {
    while (true) {
      const p = await Promise.race([ai.next(), timer]);
      if ('done' in p && p.done)
        return ai.return?.(p.value);
      if ('debounced' in p && p.debounced) {
        if (p.debounced === last)
          yield p.value;
      } else {
        // We have a new value from the src
        clearTimeout(last);
        timer = new Promise(resolve => {
          last = setTimeout(() => {
            resolve({ debounced: last, value: p.value })
          }, milliseconds);
        });
      }
    }
  } catch (ex) {
    ai.throw?.(ex);
  }
}

async function* waitFor<U>(this: AsyncIterable<U>, cb: (done: (...a: unknown[]) => void) => void): AsyncIterable<U> {
  const ai = this[Symbol.asyncIterator]();
  try {
    while (true) {
      const p = await ai.next();
      if (p.done) {
        return ai.return?.(p.value);
      }
      await new Promise(resolve => cb(resolve));
      yield p.value;
    }
  } catch (ex) {
    ai.throw?.(ex);
  }
}

async function* count<U extends {}, K extends string>(this: AsyncIterable<U>, field: K) {
  const ai = this[Symbol.asyncIterator]();
  let count = 0;
  try {
    for await (const value of this) {
      const counted = {
        ...value,
        [field]: count++
      }
      yield counted;
    }
  } catch (ex) {
    throw ex;
  }
}

function retain<U extends {}>(this: AsyncIterable<U>): AsyncIterableIterator<U> & { value: U, done: boolean } {
  const ai = this[Symbol.asyncIterator]();
  let prev: IteratorResult<U, any>;
  return {
    [Symbol.asyncIterator]() { return this },
    next() {
      const n = ai.next();
      n.then(p => prev = p);
      return n;
    },
    return(value?: any) {
      return ai.return?.(value) ?? Promise.resolve({done: true as const, value });
    },
    throw(...args: any[]) {
      return ai.throw?.(args) ?? Promise.resolve({done: true as const, value: args[0] })
    },
    get value() {
      return prev.value;
    },
    get done() {
      return Boolean(prev.done);
    }
  }
}

function broadcast<U, X>(this: AsyncIterable<U>, pipe: ((dest: AsyncIterable<U>) => AsyncIterable<X>) = (x => x as unknown as AsyncIterable<X>)): AsyncIterable<X> {
  const ai = this[Symbol.asyncIterator]();
  const b = broadcastIterator<U>(/*() => console.log("..stooped")*/);
  (function update() {
    ai.next().then(v => {
      if (v.done) {
        // Meh - we throw these away for now.
        // console.log(".broadcast done");
      } else {
        b.push(v.value);
        update();
      }
    }).catch(ex => b.close(ex));
  })();

  return {
    [Symbol.asyncIterator]() {
      const dest = pipe(b);
      return dest[Symbol.asyncIterator]();
    }
  }
}

async function consume<U>(this: AsyncIterable<U>, f: (u: U) => void): Promise<void> {
  for await (const u of this)
    f(u);
}

