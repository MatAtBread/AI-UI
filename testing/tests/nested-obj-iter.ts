/// <reference path="../test.env.d.ts"/>

import { Iterators } from '../../module/src/ai-ui';
import { IterableType } from '../../module/src/iterators';
const { defineIterableProperty } = Iterators;
type O = {
  i: {
    s: IterableType<string>;
    n: IterableType<string>;
  }
}

const res = Promise.resolve();

async function up(l: string, x: O, async: number | false | Promise<any> = false) {
  console.log(l, x.i.valueOf());
  x.i.s = '=s';
  if (async) await (async instanceof Promise ? async : Test.sleep(async));

  console.log(l, x.i.valueOf());
  x.i = { s: '=obj2', n: '=n2' }
  if (async) await (async instanceof Promise ? async : Test.sleep(async));

  console.log(l, x.i.valueOf());
  x.i.s = '=s2';
  if (async) await (async instanceof Promise ? async : Test.sleep(async));

  console.log(l, x.i.valueOf());
  x.i = { s: '=obj3', n: '=n3' }
  if (async) await (async instanceof Promise ? async : Test.sleep(async));

  console.log(l, 'done', x.i.valueOf());
}

const o = defineIterableProperty({}, 'i', { s: 's', n: 'n' });
await up('o', o);

const ao = defineIterableProperty({}, 'i', { s: 's', n: 'n' });
await up('ao', ao, res);

const p = defineIterableProperty({}, 'i', { s: 's', n: 'n' });
p.i.consume!(j => console.log('p.i', j));
p.i.s.consume!(j => console.log('p.i.s', j));
p.i.n.consume!(j => console.log('p.i.n', j));
await up('p.consumed', p);

const ap = defineIterableProperty({}, 'i', { s: 's', n: 'n' });
ap.i.consume!(j => console.log('ap.i', j));
ap.i.s.consume!(j => console.log('ap.i.s', j));
ap.i.n.consume!(j => console.log('ap.i.n', j));
await up('ap.consumed', ap, res);

await Test.sleep(2);

const m = defineIterableProperty({},'i', { s: 's', n: 'n' });
m.i.map!(x => x).consume(x => console.log('m1',x));
m.i.map!(x => x).consume(x => console.log('m2',x));
await up('mA', m, res);
await Test.sleep(2);
await up('mB', m, res);
await Test.sleep(2);
