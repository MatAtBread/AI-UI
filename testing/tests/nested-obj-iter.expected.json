[
  [1,"o",{"s":"s","n":"n"}],
  [2,"o",{"s":"=s","n":"n"}],
  [3,"o",{"s":"=obj2","n":"=n2"}],
  [4,"o",{"s":"=s2","n":"=n2"}],
  [5,"o","done",{"s":"=obj3","n":"=n3"}],
  [6,"ao",{"s":"s","n":"n"}],
  [7,"ao",{"s":"=s","n":"n"}],
  [8,"ao",{"s":"=obj2","n":"=n2"}],
  [9,"ao",{"s":"=s2","n":"=n2"}],
  [10,"ao","done",{"s":"=obj3","n":"=n3"}],
  [11,"p\\.consumed",{"s":"s","n":"n"}],
  [12,"p\\.consumed",{"s":"=s","n":"n"}],
  [13,"p\\.consumed",{"s":"=obj2","n":"=n2"}],
  [14,"p\\.consumed",{"s":"=s2","n":"=n2"}],
  [15,"p\\.consumed","done",{"s":"=obj3","n":"=n3"}],
  [16,"ap\\.consumed",{"s":"s","n":"n"}],
  [17,"ap\\.consumed",{"s":"=s","n":"n"}],
  [18,"p\\.i",{"s":"=s","n":"n"}],
  [19,"ap\\.consumed",{"s":"=obj2","n":"=n2"}],
  [20,"p\\.i\\.s","=s"],
  [21,"ap\\.i",{"s":"=s","n":"n"}],
  [22,"ap\\.consumed",{"s":"=s2","n":"=n2"}],
  [23,"p\\.i",{"s":"=s2","n":"=n2"}],
  [24,"ap\\.i\\.s","=s"],
  [25,"ap\\.consumed","done",{"s":"=obj3","n":"=n3"}],
  [26,"p\\.i\\.n","=n2"],
  [27,"ap\\.i",{"s":"=s2","n":"=n2"}],
  [28,"p\\.i\\.s","=obj3"],
  [29,"ap\\.i\\.n","=n2"],
  [30,"ap\\.i\\.s","=obj3"],
  [31,"mA",{"s":"s","n":"n"}],
  [32,"mA",{"s":"=s","n":"n"}],
  [33,"mA",{"s":"=obj2","n":"=n2"}],
  [34,"mA",{"s":"=s2","n":"=n2"}],
  [35,"m1",{"s":"=s","n":"n"}],
  [36,"m2",{"s":"=s","n":"n"}],
  [37,"mA","done",{"s":"=obj3","n":"=n3"}],
  [38,"m1",{"s":"=obj3","n":"=n3"}],
  [39,"m2",{"s":"=obj3","n":"=n3"}],
  [40,"mB",{"s":"=obj3","n":"=n3"}],
  [41,"mB",{"s":"=s","n":"=n3"}],
  [42,"mB",{"s":"=obj2","n":"=n2"}],
  [43,"mB",{"s":"=s2","n":"=n2"}],
  [44,"m1",{"s":"=s","n":"=n3"}],
  [45,"m2",{"s":"=s","n":"=n3"}],
  [46,"mB","done",{"s":"=obj3","n":"=n3"}],
  [47,"m1",{"s":"=obj3","n":"=n3"}],
  [48,"m2",{"s":"=obj3","n":"=n3"}]
]